### ISSUE TITLE
twisted.trial.unittest.TestCase.failUnlessFailure should provide more information

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#1869]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#1869|
|Type|enhancement|
|Created|2006-06-26 18:05:51Z|

Currently if the assertion fails, very little information is available to the developer about why it failed.  The type of exception which was raised is shown, but no traceback or even string version of the exception is shown.  This makes it hard to figure out why the assertion failed.

<details><summary>Searchable metadata</summary>

```
trac-id__1869 1869
type__enhancement enhancement
reporter__exarkun exarkun
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__ 
time__1151345151000000 1151345151000000
changetime__1152419655000000 1152419655000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml set owner to @exarkun|
|-|-|

Ready for review in [branches/failure-info-1869](branches/failure-info-1869)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @jml<br>@exarkun set status to `new`|
|-|-|

failUnlessFailure docstring is a mess.  Can you fix the pre-existing punctuation and spelling errors (as long as you are reformatting it anyway)?

Otherwise very nice branch.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml set status to `closed`|
|-|-|

(In [17509]) Provide stack trace information when assertFailure fails.

 * Author: jml
 * Reviewer: exarkun
 * Fixes [#1869](https://github.com/twisted/twisted/issues/1869)

assertFailure (aka failUnlessFailure) used to only report the class
of the unexpected exception. Now it also reports the brief stacktrace
and message from that exception.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
